[SandboxVec] Tag insts in a Region with metadata.

slackito opened the pull request:
For each region, we create a metadata node. Then when an instruction is added to the Region, it gets tagged with the metadata node for that region. In the following example, we have a Region that contains only the `%t0` instruction.

```
define i8 @foo(i8 %v0, i8 %v1) {
  %t0 = add i8 %v0, 1, !sbvec !0
  %t1 = add i8 %t0, %v1
  ret i8 %t1
}

!0 = distinct !{!"region"}
```

This commit also adds a function to create regions from metadata already present in a Function.

This metadata can be used for debugging: if we dump IR before a Region pass, the IR will contain enough info to re-create the Region and run the pass by itself in a later invocation.

slackito committed:
[SandboxVec] Tag insts in a Region with metadata.

For each region, we create a metadata node with the Region ID. Then
when an instruction is added to the Region, it gets tagged with the
metadata node for that region. In the following example, we have a
Region that contains only the `%t0` instruction.

```
define i8 @foo(i8 %v0, i8 %v1) {
  %t0 = add i8 %v0, 1, !sbvec !0
  %t1 = add i8 %t0, %v1
  ret i8 %t1
}

!0 = !{!"region", i32 0}
```

This commit also adds a function to create regions from metadata already
present in a Function.

This metadata can be used for debugging: if we dump IR before a Region
pass, the IR will contain enough info to re-create the Region and run
the pass by itself in a later invocation.

slackito committed:
Merge branch 'main' into region-metadata

vporpo commented on llvm/lib/Transforms/Vectorize/SandboxVectorizer/Region.cpp:
```diff
@@ -42,4 +54,27 @@ void Region::dump() const {
+    for (Instruction &Inst : BB) {
+      if (auto *MDN = cast<llvm::Instruction>(Inst.Val)->getMetadata(MDKind)) {
+        Region *R = nullptr;
+        auto It = MDNToRegion.find(MDN);
```
Perhaps move this code into a lambda?
```
auto GetOrCreateNode = [&MDNToRegion](MDNode *MDN, std::unique_ptr<Region> &Regions) {
  Region *R = nullptr;
  auto It = MDNToRegion.find(MDN);
  if (It == MDNToRegion.end()) {
     Regions.push_back(std::make_unique<Region>(Ctx));
     R = Regions.back().get();
     MDNToRegion[MDN] = R;
  } else {
     R = It->second;
  }
  return R;
};
Region *R = GetOrCreateNode(MDN, Regions);
R->add(&Inst);
```

slackito replied:
I don't think the lambda wins us any clarity here. It needs to access both `Regions` and `MDNToRegion` from the surrounding scope, so it's not very self-contained anyway, and this block is just 10 lines of code.

vporpo commented on llvm/unittests/Transforms/Vectorize/SandboxVectorizer/RegionTest.cpp:
```diff
@@ -79,3 +78,31 @@ define i8 @foo(i8 %v0, i8 %v1) {
+#ifndef NDEBUG
+  EXPECT_EQ(Rgn, *Regions[0].get());
+#endif
+}
```
The round-trip test is very useful, but it doesn't actually test that the right type of metadata is being created. I would add some checks that the metadata nodes are actually being created.

A separate test that just builds a region out of metadata IR would also be helpful as a reference of how a region looks like in IR.

tmsri replied:
+1 Another test with explicit metadata being converted to regions?

tmsri reviewed (changes requested):
More tests.

vporpo commented on llvm/lib/Transforms/Vectorize/SandboxVectorizer/Region.cpp:
```diff
@@ -13,13 +13,25 @@ namespace llvm::sandboxir {
-void Region::add(Instruction *I) { Insts.insert(I); }
+void Region::add(Instruction *I) {
+  Insts.insert(I);
+  cast<llvm::Instruction>(I->Val)->setMetadata(MDKind, RegionMDN);
```
Let's also add a TODO about a potential redesign where we tag instructions lazily, perhaps at the end of the region pass manager, to save compile time. It might be worth exploring in the future.

slackito committed:
Address review feedback

- Added new test case that creates regions from explicit metadata.
- As suggested offline by @aeubanks, removed region IDs in favor of
  using `distinct` metadata nodes to identify each region.

Also removed a stale include for "InstructionCost" left over from the
original version of PR108899. All cost-related stuff was removed.

slackito committed:
Add requested TODO

slackito committed:
Merge branch 'main' into region-metadata

slackito committed:
removed extra whitespace at end of line

vporpo commented on llvm/unittests/Transforms/Vectorize/SandboxVectorizer/RegionTest.cpp:
```diff
@@ -79,3 +78,58 @@ define i8 @foo(i8 %v0, i8 %v1) {
+  sandboxir::Region Rgn(Ctx);
+  Rgn.add(T0);
+  Rgn.add(T1);
+
```
Let's also add a check here to test that the metadata has been created. Or otherwise a separate test Region to IR that checks the IR dump? Wdyt ?

tmsri commented on llvm/unittests/Transforms/Vectorize/SandboxVectorizer/RegionTest.cpp:
```diff
@@ -79,3 +78,58 @@ define i8 @foo(i8 %v0, i8 %v1) {
+  ASSERT_EQ(1U, Regions.size());
+#ifndef NDEBUG
+  EXPECT_EQ(Rgn, *Regions[0].get());
+#endif
```
How about dumping the IR and explicitly check for required number of "distinct !{!"region"}" and create more than one region?

slackito replied:
I don't think that is any better than doing a roundtrip and checking the Regions created from metadata in the dumped IR, but done.

slackito committed:
clang-format

slackito committed:
Add new test that dumps a module and checks for the right nr of regions.

vporpo commented on llvm/unittests/Transforms/Vectorize/SandboxVectorizer/RegionTest.cpp:
```diff
@@ -106,6 +106,44 @@ define i8 @foo(i8 %v0, i8 %v1) {
+           /*IsForDebug=*/true);
+
+  // Check that the dumped IR contains two `distinct !{!"region"}` nodes.
+  std::string RegionString = "distinct !{!\"region\"}";
```
1. Maybe use a string literal to avoid escaping the double quotes `R"(distinct !{!"region"})"`
2. The IR is quite small, so why not compare the whole dump? If we change the dumps it would be relatively easy to update it.

vporpo commented on llvm/unittests/Transforms/Vectorize/SandboxVectorizer/RegionTest.cpp:
```diff
@@ -106,6 +106,44 @@ define i8 @foo(i8 %v0, i8 %v1) {
+
+  // Check that the dumped IR contains two `distinct !{!"region"}` nodes.
+  std::string RegionString = "distinct !{!\"region\"}";
+  auto Pos = output.find(RegionString);
```
I think `llvm::StringRef` implements a `contains()` function.

slackito replied:
I don't see a `contains()` variant that takes a start position, so then to find the second occurrence I should create a `substr()` and use `contains()` on that? :|

slackito replied:
Never mind, I could use `count` for that. I'm going to follow your suggestion in the other comment and compare the whole dump anyway.

slackito committed:
Merge branch 'main' into region-metadata

slackito committed:
Check whole IR dump in DumpedMetadata test case

vporpo commented on llvm/unittests/Transforms/Vectorize/SandboxVectorizer/RegionTest.cpp:
```diff
@@ -79,3 +78,102 @@ define i8 @foo(i8 %v0, i8 %v1) {
+  llvm::raw_string_ostream RSO(output);
+  M->print(RSO, nullptr, /*ShouldPreserveUseListOrder=*/true,
+           /*IsForDebug=*/true);
+
```
Please add a TODO that we should replace this with a lit test, because lit tests are designed for this type of IR comparison.

slackito replied:
Done

vporpo reviewed (approved):
Looks good to me, thanks!

slackito committed:
Add TODO

alinas reviewed (commented):
Nit: I'd suggest renaming mdnodes for clarity, in case the region name is used elsewhere.

slackito committed:
Apply suggested change from "sbvec" to "sandboxvec"

Co-authored-by: Alina Sbirlea <alina.g.simion@gmail.com>

slackito committed:
Commit suggested change from "region" to "sandboxregion"

Co-authored-by: Alina Sbirlea <alina.g.simion@gmail.com>

slackito committed:
Update test cases after suggested metadata string changes
